lib/app/dispatcher.rb

```diff
@@ -56,46 +57,74 @@ def pool
+                       else
+                         Core::SingleScheduler.new(App::Config.connector_id, POLL_INTERVAL, HEARTBEAT_INTERVAL)
+                       end
       end
 
       def start_polling_jobs!
-        scheduler.when_triggered do |connector_settings|
-          service_type = connector_settings.service_type
-          connector_id = connector_settings.id
-          index_name = connector_settings.index_name
-
-          # connector-level checks
-          unless Connectors::REGISTRY.registered?(service_type)
-            Utility::Logger.info("The service type (#{service_type}) of connector (ID: #{connector_id}) is not supported. Skipping...")
-            next
-          end
-          if index_name.nil? || index_name.empty?
-            Utility::Logger.info("The index name of connector (ID: #{connector_id}, service type: #{service_type}) is empty. Skipping...")
-            next
-          end
-          unless index_name.start_with?(Utility::Constants::CONTENT_INDEX_PREFIX)
-            Utility::Logger.info("The index name of connector (ID: #{connector_id}, service type: #{service_type}) is invalid, it must start with '#{Utility::Constants::CONTENT_INDEX_PREFIX}'. Skipping...")
-            next
-          end
-          Core::ElasticConnectorActions.ensure_content_index_exists(index_name)
-
-          pool.post do
-            send_heartbeat(connector_id, service_type)
-            Utility::Logger.info("Starting a job for connector (ID: #{connector_id}, service type: #{service_type})...")
-            job_runner = Core::SyncJobRunner.new(connector_settings, service_type)
-            job_runner.execute
-          rescue StandardError => e
-            Utility::ExceptionTracking.log_exception(e, "Job for connector (ID: #{connector_id}, service type: #{service_type}) failed due to unexpected error.")
+        scheduler.when_triggered do |connector_settings, task|
+          case task
+          when :sync
+            start_sync_task(connector_settings)
+          when :heartbeat
+            start_heartbeat_task(connector_settings)
+          when :configuration
+            start_configuration_task(connector_settings)
+          else
+            Utility::Logger.error("Unknown task type: #{task}. Skipping...")
           end
         end
       rescue StandardError => e
-        Utility::ExceptionTracking.log_exception(e, 'connector service failed due to unexpected error.')
+        Utility::ExceptionTracking.log_exception(e, 'The connector service failed due to unexpected error.')
       end
 
-      def send_heartbeat(connector_id, service_type)
-        Utility::Logger.info("Sending a heartbeat on [#{connector_id} - #{service_type}]...")
-        Core::Heartbeat.send(connector_id, service_type)
+      def start_sync_task(connector_settings)
+        # connector-level checks
+        unless Connectors::REGISTRY.registered?(connector_settings.service_type)
```

Comment:
In non-native mode, `service_type` is populated by the configuration task. If this runs before the configuration task, this will fail because `connector_settings.service_type` should be `nil`, but will eventually pass.

Reply:
That should be fixed by running everything in 1 thread, right?

Reply:
Yes you are right. ignore my comment above